packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m

```diff
@@ -543,16 +562,25 @@ - (void)setPlaybackSpeed:(double)speed {
 }
 
 - (CVPixelBufferRef)copyPixelBuffer {
+  // Ensure video sampling at regular intervals. This function is not called at exact time intervals
+  // so CACurrentMediaTime returns irregular timestamps which causes missed video frames. The range
+  // outside of which targetTime is reset should be narrow enough to make possible lag as small as
+  // possible and at the same time wide enough to avoid too frequent resets which would lead to
+  // irregular sampling. Ideally there would be a targetTimestamp of display link used by flutter
```

Comment:
Can you add more explanation on how it leads to irregular sampling? 

Can we do an experiment to always reset the time (remove the `if` check below) and see how it performs on a sample video?

Reply:
> Can you add more explanation on how it leads to irregular sampling?

Because each "reset" (meaning `if` is true) changes `targetTime` to a different value than it had. Adding the same number (`self.frameUpdater.duration`) results in regular (enough) time intervals in `targetTime` but suddenly changing it to value "outside" of that breaks that regularity. In the worst case when it changes always it is like directly using `CACurrentMediaTime`.

> Can we do an experiment to always reset the time (remove the `if` check below) and see how it performs on a sample video?

Yes I did that of course. I also calculated standard deviation and tried to compute the probability of frame drop, result was about 10x higher than I observed but I probably did something wrong in this calculation (it does not really matter). And also distribution here is not normal so it may not match calculations assuming normal distribution. Here black is the distribution of differences between results of `CACurrentMediaTime` in consecutive `copyPixelBuffer` calls (on my ios device) and green is normal distribution. Horizontal axis goes from around 8 to 24 ms with 1/60 s in middle and vertical axis is relative number of occurrences (this graph shows that there should be less drops than with normal distribution so I am satisfied enough with it as explanation why my calculated number was 10x higher):

<img width="264" alt="Screenshot 2024-11-18 at 18 45 33" src="https://github.com/user-attachments/assets/bb2a04bc-4c1c-47ab-a1cc-af27d55e3359">